Puppeteer wrapper for pdf rendering

mbayopanda opened the pull request:
This PR add puppeteer for pdf rendering as replacement of wkhtmltopdf.

mbayopanda committed:
replace wkthmltopdf by puppeteer

mbayopanda committed:
remove yarn.lock

mbayopanda committed:
pdf generator and fix report orientation

mbayopanda committed:
pdf orientation fixing

mbayopanda committed:
fix and enhancements

mbayopanda committed:
pdf service notification

jniles commented on package.json:
```diff
@@ -114,6 +114,8 @@
     "mz": "^2.7.0",
     "ngstorage": "gsklee/ngStorage#0.3.11",
+    "puppeteer": "^1.10.0",
+    "puppeteer-core": "^1.10.0",
```
Do we need both puppeteer and puppeteer-core?

jniles commented on server/lib/ReportManager.js:
```diff
@@ -236,4 +235,12 @@ class ReportManager {
+// function to encode file data to base64 encoded string
+function base64Encode(file) {
+  // read binary data
+  const bitmap = fs.readFileSync(file);
```
This might be part of the problem with the speed - we don't want to write synchronous code on the server!  Can you make this into an asynchronous operation?

jniles commented on server/lib/renderers/pdf.js:
```diff
@@ -2,96 +2,160 @@
+  args : [
+    '--no-sandbox',
+    '--disable-web-security',
+    '--disable-gpu',
```
Out of curiosity, do we actually need all these flags?  What happens if we remove them?

jniles commented on server/lib/renderers/pdf.js:
```diff
@@ -2,96 +2,160 @@
+    await page.addScriptTag({ content : content.toString() });
+  }
+
+  const pdf = await page.pdf(options);
```
Let's just `return page.pdf(options);`

jniles commented on server/lib/renderers/pdf.js:
```diff
@@ -2,96 +2,160 @@
+  for (let i = 0; i < dbStyles.length; i++) {
+    const content = dbStyles[i];
+    // eslint-disable-next-line no-await-in-loop
+    await page.addStyleTag({ content : content.toString() });
```
Since we switched from using `path:` to using `content:`, we now control the order of these scripts.  I think you can write this more efficiently by simply:
```js
await Promise.all(dbStyles);
```

Note that this will only work if you have called the `readFile` operation with 'utf8' to avoid needing to call the `.toString()` method.

jniles commented on server/lib/template/partials/enterpriseDetails.handlebars:
```diff
@@ -1,12 +1,13 @@
   <div class="col-xs-1">
-    <img src="{{absolutePath}}/{{slashed metadata.enterprise.logopath}}" width="65" height="65" alt="">
+    {{!-- <img src="{{absolutePath}}/{{slashed metadata.enterprise.logopath}}" width="65" height="65" alt=""> --}}
+    <img src="data:image/jpg;base64,{{metadata.enterprise.binaryLogo}}" width="65" height="65" alt="">
```
One thing to consider ... this is always `image/jpg`.  Should we limit all uploads to be `jpg` or should we try to convert pngs -> jpgs, or is there a way to figure out the image type ...?

mbayopanda replied:
It seems like pngs are also handled by `image/jpg`

jniles replied:
:dancer:

jniles commented on server/lib/renderers/pdf.js:
```diff
@@ -2,96 +2,160 @@
+
+function environmentStyleAndScripts() {
+  const styles = resources.cssPaths.map(cssPath => {
+    return fs.readFileSync(cssPath);
```
Can we do this in asynchronously?  Also, these should read the files as 'utf8'.

I propose that we use the `mz/fs` (which is already used in the backup scripts) to make this into a promise.  so that we are basically doing:
```js
function environmentStyleAndScripts() {
  const styles = resources.cssPaths.map(path => fs.readFile(path, 'utf8'));
  const scripts = resources.jsPaths.map(path => fs.readFile(path, 'utf8'));
  return { styles, scripts };
}
```

... then later you will be able to do:
```js
const stylesheets = await Promise.all(styles);
```

jniles commented on server/lib/renderers/pdf.js:
```diff
@@ -2,96 +2,160 @@
+const browserPromise = pptr.launch(pptrOptions)
+  .then(brwsr => {
+    // notify about the pdf service
+    debug('PDF service is running');
```
👍

jniles reviewed (commented):
@mbayopanda this is an awesome start!

We have a lot of synchronous code introduced, which will slow NodeJS down.  Let's see if we can deal with that by loading up our files asynchronously.  It's not critical for those files loaded at the start of the process, but it is better practice.

mbayopanda committed:
remove sync operations

mbayopanda committed:
fix footer template

mbayopanda committed:
skip wkhtmltopdf tests

mbayopanda committed:
remove lock files

mbayopanda committed:
update travis script

mbayopanda committed:
update dependencies by adding yarn lock file

mbayopanda committed:
close pptr browser page

mbayopanda committed:
restart pptr browser if it is closed during a pdf generation

mbayopanda committed:
install last version of yarn at Travis

mbayopanda committed:
fix style according standard

mbayopanda committed:
update yarn for travis

jniles commented on .travis.yml:
```diff
@@ -3,6 +3,8 @@ dist: trusty
 
 before_install:
   - jdk_switcher use oraclejdk8
+  - "curl -o- -L https://yarnpkg.com/install.sh | bash -s --"
```
:+1:

jniles commented on client/src/js/components/bhRenderOption.js:
```diff
@@ -26,6 +26,8 @@ function RenderOptionController() {
+    ctrl.setOrientation();
   };
+
+  ctrl.setOrientation = () => ctrl.onChange({ orientation : ctrl.orientation });
```
This is a good change.  👍

jniles commented on client/src/modules/templates/bhRenderOption.tmpl.html:
```diff
@@ -2,13 +2,13 @@
   <div>
     <label class="radio-inline">
-      <input type="radio" name="portrait" id="portrait"  value="portrait" ng-click ="$ctrl.setOrientation()" ng-model="$ctrl.orientation" required>
+      <input type="radio" name="portrait" id="portrait"  value="portrait" ng-change ="$ctrl.setOrientation()" ng-model="$ctrl.orientation" required>
```
💃

jniles commented on gulpfile.js:
```diff
@@ -32,13 +32,24 @@ const supportedLanguages = {
       'client/src/js/define.js',
       'client/src/js/app.js',
       'client/src/**/*.js',
+      '!client/src/js/plugins/*.js',
```
Do barcodes still render on the client-side in receipts with HTML format?

jniles commented on package.json:
```diff
@@ -123,8 +124,7 @@
     "uuid-parse": "^1.0.0",
     "webcam": "jonashartmann/webcam-directive#3.1.2",
-    "winston": "^3.1.0",
-    "wkhtmltopdf": "^0.3.3"
```
💃

jniles commented on package.json:
```diff
@@ -114,6 +114,7 @@
     "mysql": "^2.14.0",
     "mz": "^2.7.0",
     "ngstorage": "gsklee/ngStorage#0.3.11",
+    "puppeteer": "^1.10.0",
```
Have you tried using puppeteer-core on windows?  Does it work?

mbayopanda replied:
No, it doesn't works, it require to download chrome : 
![image](https://user-images.githubusercontent.com/5445251/52211909-7b8e4f80-288b-11e9-9bb2-90a28fd89f94.png)

`puppeteer-core` requires also the path to chromium executable : 
https://github.com/GoogleChrome/puppeteer/blob/master/docs/api.md#puppeteer-vs-puppeteer-core

jniles commented on server/lib/renderers/pdf.js:
```diff
@@ -2,96 +2,160 @@
-      return streamToPromise(pdfStream);
-    });
+  // check if browser is closed
+  browser._process.once('close', () => {
```
Looks like this might have a memory leak :)

jniles reviewed (commented):
Just a few clarification comments.  Looks like Travis is giving us trouble for some reason :/

mbayopanda committed:
fix error on puppeteer

mbayopanda committed:
update travis for yarn

mbayopanda committed:
use of npm in travis for test

jniles commented on server/lib/template/partials/enterpriseDetails.handlebars:
```diff
@@ -1,12 +1,12 @@
   <h3 style="margin: 0px;">{{metadata.enterprise.name}}</h3>
   <strong>{{metadata.project.name}} ({{metadata.project.abbr}})</strong> <br>
-  <span class="text-capitalize">{{translate 'FORM.LABELS.ADDRESS'}}</span>: {{enterprise.location}} <br>
+  <span class="text-capitalize">{{translate 'FORM.LABELS.ADDRESS'}}</span>: {{metadata.enterprise.location}} <br>
```
:dancer:

jniles reviewed (commented):
So...

It looks like my advice was wrong.  According to [this blog post](https://docs.browserless.io/blog/2018/06/04/puppeteer-best-practices.html), they don't want you to spin up a browser unless it is absolutely necessary and to close it each time.

Everything I've seen seems to indicate the same thing.  Let's see what happens if we change course and try to get it working by just rendering with the browser each time...

jniles committed:
fix(renderers/pdf): rm unnecessary async code

This commit makes a few asynchronous code portions synchronous (since
they don't need to be async).  It also launches a browser instance PER
HTTP request.  In order to maintain a chromium process in the
background, we would need to be able to listen for any exit events and
close the chromium process.  However, best practice seems to be to start
up a new browser on each request.

jniles committed:
Merge remote-tracking branch 'upstream/master' into pptr-wrapper

jniles committed:
test(renderers): unit test the PDF renderer

We needed to unit test the PDF renderer after the transition from
wkhtmltopdf to puppeteer.  The same tests have been rewritten and
updated to account for a few minor changes between the two.
Additionally, the plugins system has been dropped for directly injecting
JS into the handlebars templates by reading from node_modules.

jniles committed:
fix(test): work around for slow buffer diff

It seems diff-ing a buffer may slow mocha to a halt.  This workaround
was proposed in the mocha issues to prevent such a diff from occurring.

jniles committed:
chore(ci): do not worry about xvfb

We needed xvfb to render consistent PDFs with WKHTMLTOPDF.  However,
since we use puppeteer now, there isn't a need to use xvfb.

jniles committed:
chore(ci): test on latest node LTS

A shortcut for "latest LTS" is lts/* in NVM.  Since Travis uses nvm,
we've changed the install to the latest LTS so that we don't have to
bump the versions ever six months.

jniles committed:
test(xlsx): label test properly

The test called itself "util.js" before, but it is now
"renderers/xlsx.js" as expected.  The code has also been upgraded to
async/await instead of done() callbacks.

jniles committed:
test: account for increased viewport

The size of the screen changed, so we need to account for the larger
viewport in our tests.

jniles committed:
chore: suppress logging renderers

jniles committed:
Merge remote-tracking branch 'upstream/master' into pptr-wrapper

jniles reviewed (commented):
@mbayopanda, can you give this a test on your machine to make sure it works on a windows machine?
